meter/consolidation_registry.cc

```diff
@@ -57,41 +40,53 @@ void ConsolidationRegistry::update_from(
     const Measurements& measurements) noexcept {
   std::lock_guard<std::mutex> guard{mutex};
 
-  double new_value;
   for (const auto& m : measurements) {
-    auto& my_m = my_measures[m.id];
-    my_m.id = m.id;
-    if (my_m.timestamp == m.timestamp) {
-      // duplicate update
+    if (std::isnan(m.value)) {
       continue;
     }
-    my_m.timestamp = m.timestamp;
+
+    auto it = my_values.find(m.id);
     auto op = aggregate_op(m);
-    switch (op) {
-      case AggrOp::Add:
-        new_value = addnan(my_m.value, m.value / update_multiple);
-        my_m.value = new_value;
-        break;
-      case AggrOp::Max:
-        new_value = maxnan(m.value, my_m.value);
-        my_m.value = new_value;
-        break;
+    if (it == my_values.end()) {
+      it = my_values
+               .emplace(values_map::value_type{
+                   m.id, ConsolidatedValue(op == AggrOp::Max, update_multiple,
+                                           reporting_frequency_, clock_)})
+               .first;
     }
+    auto& my_value = it->second;
+    my_value.update(m.value);
```

Comment:
Won't the activity for the last update-interval get applied to the next reporting interval? Consider `update_frequency_ = 5s` and `reporting_frequency_ = 10s`, I think the update pattern would look something like:

```
|-----|-----|-----|-----|-----|-----|  5s step numbers
 3 4    6  7   2    1  3  5     2  1

|----- -----|----- -----|----- -----|----- -----|  10 consolidated number
       7     13    2     4     5     3
```

That is, the consolidated registry would see the aggregated update after the finer grained one had completed. So on the boundary, when the last 5s interval is done so is the consolidated interval, so it will get applied to the next interval rather than the one where the activity took place. Is that correct or am I missing something?

Should we try to preserve the time window for the consolidation?

Reply:
You're right. I was thinking that we'd be up to 5s late which is more or less the behavior we'll inherit when we start sending the metrics through a thin client. What do you think we should do?

Reply:
That is probably fine, just wanted to double check my understanding is correct. For the java implementation I will probably try to avoid this as it would be used for the aggregator and thus it would be additive to the delay to send with the thin client.